pydrofoil/real.py

```diff
@@ -0,0 +1,99 @@
+import sys
+from rpython.rlib.rbigint import rbigint, _divrem as bigint_divrem
+from rpython.rlib.rarithmetic import r_uint, intmask, string_to_int, ovfcheck, \
+        int_c_div, int_c_mod, r_ulonglong
+
+class Real(object):
+    def __init__(self, num, den):
+        self.num = num
+        self.den = den
+
+    @staticmethod
+    def fromint(num, den=1):
+        sign = num/abs(num)*den/abs(den)
+        num = rbigint.fromint(abs(num))
+        den = rbigint.fromint(abs(den)*sign)
+        # assert den.int_ne(0)
+        assert not den.int_eq(0), "denominator cannot be 0"
+        return Real(num.div(num.gcd(den)), den.div(num.gcd(den)))
```

Comment:
the code that reduces the fraction occurs in every single call to `Real` in this file. maybe you should do the reduction a single time, in the `__init__`. Also, you compute the expression `num.gcd(den)` twice, which is inefficient.

Reply:
yes, I would change it.